src/houghcircle.jl

```diff
@@ -0,0 +1,134 @@
+    min_radius   = minimum circle radius
+    max_radius   = maximum circle radius
+"""
+
+function hough_circle_gradient{T<:Integer}(
+        img::AbstractArray{T,2},
+        scale::Number, min_dist::Number,
+        canny_thres::Number, vote_thres::Number,
+        min_radius::Integer, max_radius::Integer)
+
+    rows,cols=size(img)
+
+    img_edges = canny(img, 1, canny_thres, canny_thres/4, percentile=false)
+    dx, dy=imgradients(img, KernelFactors.ando3)
+    _, img_phase = magnitude_phase(dx, dy)
+
+    non_zeros=CartesianIndex{2}[]
+    centers=CartesianIndex{2}[]
+    circle_centers=CartesianIndex{2}[]
+    circle_radius=Integer[]
+    votes=zeros(Integer, Int(floor(rows/scale))+1, Int(floor(cols/scale))+1)
+
+    f = CartesianIndex(map(r->first(r), indices(votes)))
+    l = CartesianIndex(map(r->last(r), indices(votes)))
+
+    function vote!(img, x, y)
+        i=Int(floor(x))+1
+        j=Int(floor(y))+1
+        p=CartesianIndex(i,j)
+
+        if min(f,p)==f && max(l,p)==l
+            votes[p]+=1
+        end
+
+        i=Int(floor(x))+2
+        j=Int(floor(y))+1
+        p=CartesianIndex(i,j)
+
+        if min(f,p)==f && max(l,p)==l
+            votes[p]+=1
+        end
+
+        i=Int(floor(x))+1
+        j=Int(floor(y))+2
+        p=CartesianIndex(i,j)
+
+        if min(f,p)==f && max(l,p)==l
+            votes[p]+=1
+        end
+
+        i=Int(floor(x))+2
+        j=Int(floor(y))+2
+        p=CartesianIndex(i,j)
+
+        if min(f,p)==f && max(l,p)==l
+            votes[p]+=1
+        end
+    end
+
+    for i in indices(img, 1)
+        for j in indices(img, 2)
+            if img_edges[i,j]!=0
+                sin_theta = -cos(img_phase[i,j]);
+                cos_theta = sin(img_phase[i,j]);
+
+                for r in min_radius:max_radius
+                    x=(i+r*sin_theta)/scale
+                    y=(j+r*cos_theta)/scale
+                    vote!(img, x, y)
+
+                    x=(i-r*sin_theta)/scale
+                    y=(j-r*cos_theta)/scale
+                    vote!(img, x, y)
+                end
+                push!(non_zeros, CartesianIndex{2}(i,j));
+            end
+        end
+    end
+
+    for i in findlocalmaxima(votes)
+        if votes[i]>vote_thres
+            push!(centers, i);
+        end
+    end
+
+    sort!(centers, lt=(a, b) -> votes[a]>votes[b])
+
+    dist(a, b) = sqrt(sum(abs2, (a-b).I))
+    votes=Array(Integer, Int(floor(dist(f,l))+1))
+
+    for center in centers
+        center=(center-1)*scale
+        fill!(votes, 0)
+
+        too_close=false
+        for circle_center in circle_centers
+            if dist(center, circle_center)< min_dist
+                too_close=true
+                break
+            end
+        end
+        if too_close==true
+            continue;
+        end
+
+        for point in non_zeros
+            votes[Int(floor(dist(center, point)/scale))+1]+=1
+        end
+
+        voters, radius = findmax(votes)
```

Comment:
If I understand correctly, this will mean that if you have concentric circles, only one will be detected?